frame/support/procedural/src/construct_runtime/expand/metadata.rs

```diff
@@ -0,0 +1,187 @@
+					.pallet_parts()
+					.iter()
+					.filter(|part| part.name() != "Pallet")
+					.map(|part| part.name())
+					.collect();
+				(pallet_declaration, filtered_names)
+			})
+		})
+		.map(|(decl, filtered_names)| {
+			let name = &decl.name;
+			let index = &decl.index;
+			let storage = expand_pallet_metadata_storage(&filtered_names, runtime, scrate, decl);
+			let calls = expand_pallet_metadata_calls(&filtered_names, runtime, scrate, decl);
+			let event = expand_pallet_metadata_events(&filtered_names, runtime, scrate, decl);
+			let constants = expand_pallet_metadata_constants(runtime, scrate, decl);
+			let errors = expand_pallet_metadata_errors(runtime, scrate, decl);
+
+			quote!{
+				#scrate::metadata::ModuleMetadata {
+					name: #scrate::metadata::DecodeDifferent::Encode(stringify!(#name)),
+					index: #index,
+					storage: #storage,
+					calls: #calls,
+					event: #event,
+					constants: #constants,
+					errors: #errors,
+				}
+			}
+		})
+		.collect::<Vec<_>>();
+
+	quote!{
+		impl #runtime {
+			pub fn metadata() -> #scrate::metadata::RuntimeMetadataPrefixed {
+				#scrate::metadata::RuntimeMetadataLastVersion {
+					modules: #scrate::metadata::DecodeDifferent::Encode(&[ #(#modules),* ]),
+					extrinsic: #scrate::metadata::ExtrinsicMetadata {
+						version: <#extrinsic as #scrate::sp_runtime::traits::ExtrinsicMetadata>::VERSION,
+						signed_extensions: <
+								<
+									#extrinsic as #scrate::sp_runtime::traits::ExtrinsicMetadata
+								>::SignedExtensions as #scrate::sp_runtime::traits::SignedExtension
+							>::identifier()
+								.into_iter()
+								.map(#scrate::metadata::DecodeDifferent::Encode)
+								.collect(),
+					},
+				}.into()
+			}
+		}
+	}
+}
+
+fn expand_pallet_metadata_storage(
+	filtered_names: &[&'static str],
+	runtime: &Ident,
+	scrate: &TokenStream,
+	decl: &Pallet,
+) -> TokenStream {
+	if filtered_names.contains(&"Storage") {
+		let instance = decl.instance.as_ref().into_iter();
+		let path = &decl.pallet;
+
+		quote!{
+			Some(#scrate::metadata::DecodeDifferent::Encode(
+				#scrate::metadata::FnEncode(
+					#path::Pallet::<#runtime #(, #path::#instance)*>::storage_metadata
+				)
+			))
+		}
+	} else {
+		quote!(None)
+	}
+}
+
+fn expand_pallet_metadata_calls(
+	filtered_names: &[&'static str],
+	runtime: &Ident,
+	scrate: &TokenStream,
+	decl: &Pallet,
+) -> TokenStream {
+	if filtered_names.contains(&"Call") {
+		let instance = decl.instance.as_ref().into_iter();
+		let path = &decl.pallet;
+
+		quote!{
+			Some(#scrate::metadata::DecodeDifferent::Encode(
+				#scrate::metadata::FnEncode(
+					#path::Pallet::<#runtime #(, #path::#instance)*>::call_functions
+				)
+			))
+		}
+	} else {
+		quote!(None)
+	}
+}
+
+fn expand_pallet_metadata_events(
+	filtered_names: &[&'static str],
+	runtime: &Ident,
+	scrate: &TokenStream,
+	decl: &Pallet,
+) -> TokenStream {
+	if filtered_names.contains(&"Event") {
+		let mod_name = &decl.pallet.inner.segments.last().unwrap().ident;
```

Comment:
This is probably not the best way to name modules specified with a path. I am thinking that a better way would be something like `nested_module` for a module that's located in `nested::module`. Doing so would allow `another::module` and `nested::module` to coexist in `construct_runtime!`.